contrib/docker/builder/Dockerfile

```diff
@@ -8,8 +8,23 @@ RUN add-apt-repository ppa:bitcoin/bitcoin
 RUN apt-get update && apt-get install -y \
     autoconf automake bsdmainutils ccache cmake curl g++ g++-mingw-w64-x86-64 gcc gcc-mingw-w64-x86-64 git \
     libboost-all-dev libbz2-dev libcap-dev libdb4.8-dev libdb4.8++-dev libevent-dev libminiupnpc-dev libprotobuf-dev \
-    libqrencode-dev libssl-dev libtool libzmq3-dev make pkg-config protobuf-compiler python-pip qtbase5-dev \
-    qttools5-dev-tools python3-zmq
+    libqrencode-dev libssl-dev libtool make pkg-config protobuf-compiler python-pip qtbase5-dev \
+    qttools5-dev-tools python3-zmq zlib1g-dev build-essential
+
+
+# Install minizip from source (unavailable from apt)
+RUN curl -L https://www.zlib.net/zlib-1.2.11.tar.gz | tar -xz -C /tmp && \
+    cd /tmp/zlib-1.2.11/contrib/minizip && \
+    autoreconf -fi && \
+    ./configure --enable-shared=no --with-pic && \
+    make -j$(nproc) install && \
+    cd / && rm -rf /tmp/zlib-1.2.11
+
+# Install zmq from source (outdated version on apt)
+RUN curl -L https://github.com/zeromq/zeromq4-1/releases/download/v4.1.4/zeromq-4.1.4.tar.gz | tar -xz -C /tmp && \
```

Comment:
Sorry just notice there is a problem with this. We need some way to install this version alongside with the version provided in official repository. The reason is if it install together it might break other packages. Even we did not explicitly install the version in the official repository it may implicitly install if we install a package that required it.

Reply:
The version being installed in `depends/packages` is ZMQ 4.3.1. I can install this version?

Reply:
official repository version depends on which ubuntu version is being used to install. Previously we installing `libzmq3-dev` here, but the version on apt in Ubuntu 14.04 is outdated.

Reply:
I mean we need some way to install this version alongside with the outdated version that provided in the official repository. I can give you some guide on this. If you need my guide please provide all of the file path installed by both outdated version and this version so we can know what steps we need to do.

Reply:
Not quite sure what you mean as only one version can be used for compilation. Newer versions of ZMQ have support for all the existing functions.
To install outdated version:
`sudo apt-get install libzmq3-dev`
Any version of ZMQ can be installed from source by the instructions in the Dockerfile.

Reply:
The other packages that depend on ZeroMQ will trigger the installation of the outdated version due to it don't know about the version that we installed manually. There is a high chance that this will conflicting with the manually installed one. So we need to make sure the manually installed one will not conflict with the official one.

Reply:
It won't conflict because the official version is not getting installed, I removed the installation of the official one. One thing to note also is that ZMQ is not actually being used directly in the project until now - before it was just an example. 
Is there a way we can test the new Dockerfile on a Zcoin build?

Reply:
Yes we did not explicitly install it but it will install automatically if we install the packages that required it, which we don't know which packages that required it, it may required directly or indirectly by its dependencies.

Reply:
Ok, how can we test that.

Reply:
The first step we should try is take a looks all installed files by official package and the manual one so we know what to do next.

Reply:
@ultimaweapon All files from current repository that depend on `libzmq` are in `src/zmq`. Newer zmq from source work with these files. And no packages depend on either version of `libzmq`. In `client-api` repository, build works with new `libzmq`. So I think ok to proceed with this.

Reply:
Tested full zcoin build in the image from this Dockerfile on `master` and `clientapi` branches, both compiled without issues.

Reply:
Sorry I mean OS packages that installing from `apt-get`, not our dependencies package.

Reply:
Ok, well also in that case there does not seem to be any issues with installation.

Reply:
checked, apt installs zmq3 and zmq3-dev to
```
zmq3-dev
/usr
/usr/include
/usr/include/zmq.hpp
/usr/include/zmq_utils.h
/usr/include/zmq.h
/usr/lib
/usr/lib/x86_64-linux-gnu
/usr/lib/x86_64-linux-gnu/pkgconfig
/usr/lib/x86_64-linux-gnu/pkgconfig/libzmq.pc
/usr/lib/x86_64-linux-gnu/libzmq.a
/usr/share
/usr/share/doc
/usr/share/doc/libzmq3-dev
/usr/share/doc/libzmq3-dev/copyright
/usr/share/man
/usr/share/man/man7
/usr/share/man/man7/zmq_null.7.gz
/usr/share/man/man7/zmq_tcp.7.gz
/usr/share/man/man7/zmq.7.gz
/usr/share/man/man7/zmq_epgm.7.gz
/usr/share/man/man7/zmq_curve.7.gz
/usr/share/man/man7/zmq_plain.7.gz
/usr/share/man/man7/zmq_ipc.7.gz
/usr/share/man/man7/zmq_inproc.7.gz
/usr/share/man/man7/zmq_pgm.7.gz
/usr/share/man/man3
/usr/share/man/man3/zmq_msg_close.3.gz
/usr/share/man/man3/zmq_unbind.3.gz
/usr/share/man/man3/zmq_init.3.gz
/usr/share/man/man3/zmq_socket_monitor.3.gz
/usr/share/man/man3/zmq_proxy.3.gz
/usr/share/man/man3/zmq_z85_decode.3.gz
/usr/share/man/man3/zmq_msg_size.3.gz
/usr/share/man/man3/zmq_disconnect.3.gz
/usr/share/man/man3/zmq_recvmsg.3.gz
/usr/share/man/man3/zmq_term.3.gz
/usr/share/man/man3/zmq_msg_init_size.3.gz
/usr/share/man/man3/zmq_sendmsg.3.gz
/usr/share/man/man3/zmq_errno.3.gz
/usr/share/man/man3/zmq_send.3.gz
/usr/share/man/man3/zmq_msg_init.3.gz
/usr/share/man/man3/zmq_msg_recv.3.gz
/usr/share/man/man3/zmq_msg_data.3.gz
/usr/share/man/man3/zmq_send_const.3.gz
/usr/share/man/man3/zmq_msg_set.3.gz
/usr/share/man/man3/zmq_msg_move.3.gz
/usr/share/man/man3/zmq_msg_init_data.3.gz
/usr/share/man/man3/zmq_recv.3.gz
/usr/share/man/man3/zmq_connect.3.gz
/usr/share/man/man3/zmq_ctx_term.3.gz
/usr/share/man/man3/zmq_socket.3.gz
/usr/share/man/man3/zmq_getsockopt.3.gz
/usr/share/man/man3/zmq_z85_encode.3.gz
/usr/share/man/man3/zmq_strerror.3.gz
/usr/share/man/man3/zmq_ctx_get.3.gz
/usr/share/man/man3/zmq_msg_send.3.gz
/usr/share/man/man3/zmq_setsockopt.3.gz
/usr/share/man/man3/zmq_msg_more.3.gz
/usr/share/man/man3/zmq_ctx_set.3.gz
/usr/share/man/man3/zmq_version.3.gz
/usr/share/man/man3/zmq_bind.3.gz
/usr/share/man/man3/zmq_msg_copy.3.gz
/usr/share/man/man3/zmq_ctx_new.3.gz
/usr/share/man/man3/zmq_ctx_destroy.3.gz
/usr/share/man/man3/zmq_msg_get.3.gz
/usr/share/man/man3/zmq_curve_keypair.3.gz
/usr/share/man/man3/zmq_close.3.gz
/usr/share/man/man3/zmq_poll.3.gz
/usr/lib/x86_64-linux-gnu/libzmq.so
/usr/share/doc/libzmq3-dev/changelog.Debian.gz

zmq3
/usr
/usr/lib
/usr/lib/x86_64-linux-gnu
/usr/lib/x86_64-linux-gnu/libzmq.so.3.1.0
/usr/share
/usr/share/doc
/usr/share/doc/libzmq3
/usr/share/doc/libzmq3/NEWS.gz
/usr/share/doc/libzmq3/AUTHORS.gz
/usr/share/doc/libzmq3/copyright
/usr/share/doc/libzmq3/changelog.Debian.gz
/usr/lib/x86_64-linux-gnu/libzmq.so.3
```

And all files which is installed manually will be install to `/usr/local/lib`. Then it could not config.

Reply:
installation log
```
Making install in doc
make[1]: Entering directory `/tmp/zeromq-4.3.1/doc'
make[2]: Entering directory `/tmp/zeromq-4.3.1/doc'
make[2]: Nothing to be done for `install-exec-am'.
 /bin/mkdir -p '/usr/local/share/man/man3'
 /bin/mkdir -p '/usr/local/share/man/man7'
 /usr/bin/install -c -m 644 zmq.7 zmq_tcp.7 zmq_pgm.7 zmq_inproc.7 zmq_ipc.7 zmq_null.7 zmq_plain.7 zmq_curve.7 zmq_tipc.7 zmq_vmci.7 zmq_udp.7 zmq_gssapi.7 '/usr/local/share/man/man7'
 /usr/bin/install -c -m 644 zmq_bind.3 zmq_unbind.3 zmq_connect.3 zmq_disconnect.3 zmq_close.3 zmq_ctx_new.3 zmq_ctx_term.3 zmq_ctx_get.3 zmq_ctx_set.3 zmq_ctx_shutdown.3 zmq_msg_init.3 zmq_msg_init_data.3 zmq_msg_init_size.3 zmq_msg_move.3 zmq_msg_copy.3 zmq_msg_size.3 zmq_msg_data.3 zmq_msg_close.3 zmq_msg_send.3 zmq_msg_recv.3 zmq_msg_routing_id.3 zmq_msg_set_routing_id.3 zmq_send.3 zmq_recv.3 zmq_send_const.3 zmq_msg_get.3 zmq_msg_set.3 zmq_msg_more.3 zmq_msg_gets.3 zmq_getsockopt.3 zmq_setsockopt.3 zmq_socket.3 zmq_socket_monitor.3 zmq_poll.3 zmq_errno.3 zmq_strerror.3 zmq_version.3 zmq_sendmsg.3 zmq_recvmsg.3 zmq_proxy.3 '/usr/local/share/man/man3'
 /usr/bin/install -c -m 644 zmq_proxy_steerable.3 zmq_z85_encode.3 zmq_z85_decode.3 zmq_curve_keypair.3 zmq_curve_public.3 zmq_has.3 zmq_timers.3 zmq_poller.3 zmq_atomic_counter_new.3 zmq_atomic_counter_set.3 zmq_atomic_counter_inc.3 zmq_atomic_counter_dec.3 zmq_atomic_counter_value.3 zmq_atomic_counter_destroy.3 '/usr/local/share/man/man3'
make[2]: Leaving directory `/tmp/zeromq-4.3.1/doc'
make[1]: Leaving directory `/tmp/zeromq-4.3.1/doc'
make[1]: Entering directory `/tmp/zeromq-4.3.1'
  CC       external/unity/unity.o
  CXX      src/src_libzmq_la-address.lo
  CXX      src/src_libzmq_la-client.lo
  CXX      src/src_libzmq_la-clock.lo
  CXX      src/src_libzmq_la-ctx.lo
  CXX      src/src_libzmq_la-curve_mechanism_base.lo
  CXX      src/src_libzmq_la-curve_client.lo
  CXX      src/src_libzmq_la-curve_server.lo
  CXX      src/src_libzmq_la-dealer.lo
  CXX      src/src_libzmq_la-devpoll.lo
  CXX      src/src_libzmq_la-dgram.lo
  CXX      src/src_libzmq_la-dish.lo
  CXX      src/src_libzmq_la-dist.lo
  CXX      src/src_libzmq_la-epoll.lo
  CXX      src/src_libzmq_la-err.lo
  CXX      src/src_libzmq_la-fq.lo
  CXX      src/src_libzmq_la-gather.lo
  CXX      src/src_libzmq_la-gssapi_mechanism_base.lo
  CXX      src/src_libzmq_la-gssapi_client.lo
  CXX      src/src_libzmq_la-gssapi_server.lo
  CXX      src/src_libzmq_la-io_object.lo
  CXX      src/src_libzmq_la-io_thread.lo
  CXX      src/src_libzmq_la-ip.lo
  CXX      src/src_libzmq_la-ip_resolver.lo
  CXX      src/src_libzmq_la-ipc_address.lo
  CXX      src/src_libzmq_la-ipc_connecter.lo
  CXX      src/src_libzmq_la-ipc_listener.lo
  CXX      src/src_libzmq_la-kqueue.lo
  CXX      src/src_libzmq_la-lb.lo
  CXX      src/src_libzmq_la-mailbox.lo
  CXX      src/src_libzmq_la-mailbox_safe.lo
  CXX      src/src_libzmq_la-mechanism.lo
  CXX      src/src_libzmq_la-mechanism_base.lo
  CXX      src/src_libzmq_la-metadata.lo
  CXX      src/src_libzmq_la-msg.lo
  CXX      src/src_libzmq_la-mtrie.lo
  CXX      src/src_libzmq_la-norm_engine.lo
  CXX      src/src_libzmq_la-null_mechanism.lo
  CXX      src/src_libzmq_la-object.lo
  CXX      src/src_libzmq_la-options.lo
  CXX      src/src_libzmq_la-own.lo
  CXX      src/src_libzmq_la-pair.lo
  CXX      src/src_libzmq_la-pgm_receiver.lo
  CXX      src/src_libzmq_la-pgm_sender.lo
  CXX      src/src_libzmq_la-pgm_socket.lo
  CXX      src/src_libzmq_la-pipe.lo
  CXX      src/src_libzmq_la-plain_client.lo
  CXX      src/src_libzmq_la-plain_server.lo
  CXX      src/src_libzmq_la-poll.lo
  CXX      src/src_libzmq_la-poller_base.lo
  CXX      src/src_libzmq_la-polling_util.lo
  CXX      src/src_libzmq_la-pollset.lo
  CXX      src/src_libzmq_la-precompiled.lo
  CXX      src/src_libzmq_la-proxy.lo
  CXX      src/src_libzmq_la-pub.lo
  CXX      src/src_libzmq_la-pull.lo
  CXX      src/src_libzmq_la-push.lo
  CXX      src/src_libzmq_la-radio.lo
  CXX      src/src_libzmq_la-radix_tree.lo
  CXX      src/src_libzmq_la-random.lo
  CXX      src/src_libzmq_la-raw_decoder.lo
  CXX      src/src_libzmq_la-raw_encoder.lo
  CXX      src/src_libzmq_la-reaper.lo
  CXX      src/src_libzmq_la-rep.lo
  CXX      src/src_libzmq_la-req.lo
  CXX      src/src_libzmq_la-router.lo
  CXX      src/src_libzmq_la-scatter.lo
  CXX      src/src_libzmq_la-select.lo
  CXX      src/src_libzmq_la-server.lo
  CXX      src/src_libzmq_la-session_base.lo
  CXX      src/src_libzmq_la-signaler.lo
  CXX      src/src_libzmq_la-socket_base.lo
  CXX      src/src_libzmq_la-socks.lo
  CXX      src/src_libzmq_la-socks_connecter.lo
  CXX      src/src_libzmq_la-stream.lo
  CXX      src/src_libzmq_la-stream_engine.lo
  CXX      src/src_libzmq_la-sub.lo
  CXX      src/src_libzmq_la-tcp.lo
  CXX      src/src_libzmq_la-tcp_address.lo
  CXX      src/src_libzmq_la-tcp_connecter.lo
  CXX      src/src_libzmq_la-tcp_listener.lo
  CXX      src/src_libzmq_la-thread.lo
  CXX      src/src_libzmq_la-timers.lo
  CXX      src/src_libzmq_la-tipc_address.lo
  CXX      src/src_libzmq_la-tipc_connecter.lo
  CXX      src/src_libzmq_la-tipc_listener.lo
  CXX      src/src_libzmq_la-trie.lo
  CXX      src/src_libzmq_la-udp_address.lo
  CXX      src/src_libzmq_la-udp_engine.lo
  CXX      src/src_libzmq_la-v1_decoder.lo
  CXX      src/src_libzmq_la-v2_decoder.lo
  CXX      src/src_libzmq_la-v1_encoder.lo
  CXX      src/src_libzmq_la-v2_encoder.lo
  CXX      src/src_libzmq_la-vmci.lo
  CXX      src/src_libzmq_la-vmci_address.lo
  CXX      src/src_libzmq_la-vmci_connecter.lo
  CXX      src/src_libzmq_la-vmci_listener.lo
  CXX      src/src_libzmq_la-xpub.lo
  CXX      src/src_libzmq_la-xsub.lo
  CXX      src/src_libzmq_la-zmq.lo
  CXX      src/src_libzmq_la-zmq_utils.lo
  CXX      src/src_libzmq_la-decoder_allocators.lo
  CXX      src/src_libzmq_la-socket_poller.lo
  CXX      src/src_libzmq_la-zap_client.lo
  CC       src/src_libzmq_la-tweetnacl.lo
  CXX      tools/curve_keygen.o
  CXX      perf/local_lat.o
  CXX      perf/remote_lat.o
  CXX      perf/local_thr.o
  CXX      perf/remote_thr.o
  CXX      perf/inproc_lat.o
  CXX      perf/inproc_thr.o
  CXX      perf/perf_benchmark_radix_tree-benchmark_radix_tree.o
  AR       external/unity/libunity.a
  CXXLD    src/libzmq.la
  CXXLD    tools/curve_keygen
  CXXLD    perf/local_lat
  CXXLD    perf/remote_lat
  CXXLD    perf/local_thr
  CXXLD    perf/remote_thr
  CXXLD    perf/inproc_lat
  CXXLD    perf/inproc_thr
  CXXLD    perf/benchmark_radix_tree
make[2]: Entering directory `/tmp/zeromq-4.3.1'
 /bin/mkdir -p '/usr/local/include'
 /bin/mkdir -p '/usr/local/lib/pkgconfig'
 /usr/bin/install -c -m 644 include/zmq.h include/zmq_utils.h '/usr/local/include'
 /bin/mkdir -p '/usr/local/lib'
 /bin/bash ./libtool   --mode=install /usr/bin/install -c   src/libzmq.la '/usr/local/lib'
 /usr/bin/install -c -m 644 src/libzmq.pc '/usr/local/lib/pkgconfig'
libtool: install: /usr/bin/install -c src/.libs/libzmq.so.5.2.1 /usr/local/lib/libzmq.so.5.2.1
libtool: install: (cd /usr/local/lib && { ln -s -f libzmq.so.5.2.1 libzmq.so.5 || { rm -f libzmq.so.5 && ln -s libzmq.so.5.2.1 libzmq.so.5; }; })
libtool: install: (cd /usr/local/lib && { ln -s -f libzmq.so.5.2.1 libzmq.so || { rm -f libzmq.so && ln -s libzmq.so.5.2.1 libzmq.so; }; })
libtool: install: /usr/bin/install -c src/.libs/libzmq.lai /usr/local/lib/libzmq.la
libtool: install: /usr/bin/install -c src/.libs/libzmq.a /usr/local/lib/libzmq.a
libtool: install: chmod 644 /usr/local/lib/libzmq.a
libtool: install: ranlib /usr/local/lib/libzmq.a
libtool: finish: PATH="/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/sbin" ldconfig -n /usr/local/lib
----------------------------------------------------------------------
Libraries have been installed in:
   /usr/local/lib

If you ever happen to want to link against installed libraries
in a given directory, LIBDIR, you must either use libtool, and
specify the full pathname of the library, or use the `-LLIBDIR'
flag during linking and do at least one of the following:
   - add LIBDIR to the `LD_LIBRARY_PATH' environment variable
     during execution
   - add LIBDIR to the `LD_RUN_PATH' environment variable
     during linking
   - use the `-Wl,-rpath -Wl,LIBDIR' linker flag
   - have your system administrator add LIBDIR to `/etc/ld.so.conf'

See any operating system documentation about shared libraries for
more information, such as the ld(1) and ld.so(8) manual pages.
----------------------------------------------------------------------
 /bin/mkdir -p '/usr/local/bin'
  /bin/bash ./libtool   --mode=install /usr/bin/install -c tools/curve_keygen '/usr/local/bin'
libtool: install: /usr/bin/install -c tools/.libs/curve_keygen /usr/local/bin/curve_keygen
make[2]: Leaving directory `/tmp/zeromq-4.3.1'
make[1]: Leaving directory `/tmp/zeromq-4.3.1'
```

Reply:
Seem like official `libzmq3` will be installed because it's dependency of `python3-zmq`. We also need to make sure our manually installation is used on CI instead of official `libzmq3`.

Reply:
Installation of official version of `libzmq3-dev` has been removed. Do you mean that `python3-zmq` will install this when it being installed? If that so, we should install this package after install manual `zmq`, so it uses manual

Reply:
I mean we should make sure manual version will be use when we make zcoin.

Reply:
I have created an image from the builder Dockerfile, accessed the image, and built the `client-api` branch, without failure. If the official version was being used this would have failed. So manual version is being used

Reply:
It should be fine. I also have tried to build same step as CI on branch `client-api` seem like it would be failed when `exodus` and `client-api` are enabled together. The issue is not related to this PR, but please check.